### PR TITLE
Week 3 Homework

### PR DESCRIPTION
**Pull request by ghost**
Haven't done exercise 1 yet.

### CODE REVIEW

File: Week3/homework-midy/guessMore.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * be able to see the changes that has
+ * been made to 'x' only if we log the function
+ * --> console.log(f1(x));
+ * That's because the change made to 'x'
+ * is just temporary.
+ * ----------------------------------
+ * As for f2, the changes made to an
+ * object are not temporary, since we
+ * are reassigning the value 'x' of 'y'.
+ */
```
Comment (nikoskleidis) [inline Week3/homework-midy/guessMore.js:28]: You are right @MidyanHamdoun ! The reason is x is a primitive number and cannot be changed that way. `y` on the other hand is an object and we can change its contents. So we always need to be careful with mutating the parameters of a function when we are dealing with objects. Maybe the caller doesn't want to change the object they send to the function

File: Week3/homework-midy/guessTheOutput.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * I think the ouput of this snippet
+ * is going to be 12, since
+ * variables are not hoisted in
+ * compile-time. So, when the
+ * code is ran,'a' will change to 12'=
+ * and 'x' will be ran after declaring 'a'
+ * (since 'x' is not a function but a variable)
+ * so when 'x' is ran, 'a' is reassigned to 12
+ * and that's why 'x' will output 12
+ */
```
Comment (nikoskleidis) [inline Week3/homework-midy/guessTheOutput.js:22]: x is assigned the function at line 5 as soon as the code gets executed. The outer function get's called as soon as the code runs. It is called IIFE (https://developer.mozilla.org/en-US/docs/Glossary/IIFE)
So at this point the value of `a` has already be assigned the value 12. When x is called the value of a is 12 so this is what gets printed

File: Week3/homework-midy/lottery.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+function threeFive(startIndex, endIndex, three, five) {
+  const numbers = [];
+
+  const startParsed = parseInt(startIndex, 10);
```
Comment (nikoskleidis) [inline Week3/homework-midy/lottery.js:4]: No need to parseInt the input here. We assume that we trust the caller

File: Week3/homework-midy/lottery.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  const endParsed = parseInt(endIndex, 10);
+
+  const random = (start, end) => {
+    for (let i = 0; i < 6; i += 1) {
+      numbers.push(Math.floor(Math.random() * (end - start + 1) + start));
+    }
+    return numbers;
+  };
```
Comment (nikoskleidis) [inline Week3/homework-midy/lottery.js:12]: Nice random number generator function. But you didn't need to generate random numbers according to the exercise. 

> The function should first generate an array containing values from start value to end value (inclusive).

So if you call it `threeFive(10, 15, sayThree, sayFive);` it should generate an array of `[10,11,12,13,14,15]`

File: Week3/homework-midy/lottery.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  random(startParsed, endParsed);
+  numbers.forEach(number => {
+    if (number % 3 === 0 && number % 5 === 0) {
+      console.log(`${number} can be divided by both 3 and 5`);
```
Comment (nikoskleidis) [inline Week3/homework-midy/lottery.js]: According to the exercise you shouldn't console.log here
> Both functions should be called if the array value is divisible by both 3 and 5.

File: Week3/homework-midy/lottery.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+function sayThree(num) {
+  console.log(`${num} can be divided by 3`);
+}
+
+function sayFive(num) {
+  console.log(`${num} can be divided by 5`);
+}
+
+console.log(threeFive(2, 20, sayThree, sayFive));
```
Comment (nikoskleidis) [inline Week3/homework-midy/lottery.js:36]: Very good solution @MidyanHamdoun!

File: Week3/homework-midy/removeDuplicates.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+const letters = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c', 'b'];
+
+function removeDuplicates(array) {
+  return [...new Set(array)];
+}
```
Comment (nikoskleidis) [inline Week3/homework-midy/removeDuplicates.js:5]: Cool way that gets the job done. If this was a real project I would say kudos and accept this solution. 
But for the purpose of this week's exercise you need to achieve the same using closures. Similar to what we did in the class

**Review by nikoskleidis — CHANGES_REQUESTED** [review]
Very good job so far. Try to do exercise 1 and also redo the removeDuplicates as requested

File: Week3/homework-midy/lottery.js (hunk 4 above)
Comment (ghost) [reply Week3/homework-midy/lottery.js:12]: Oh sorry. I thought I was supposed to generate a random number within the range of the numbers given.

File: Week3/homework-midy/removeDuplicates.js (hunk 7 above)
Comment (ghost) [reply Week3/homework-midy/removeDuplicates.js:5]: Alright I'll try that!

File: Week3/homework-midy/lottery.js (hunk 6 above)
Comment (ghost) [reply Week3/homework-midy/lottery.js:36]: Thank you!

File: Week3/homework-midy/lottery.js (hunk 5 above)
Comment (ghost) [reply Week3/homework-midy/lottery.js]: Oops.

**Pull request merged**